R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+  
+  nb <- rowSums(nbk)
+  rho <- c(sum(ws * yobs * (1-zobs)) / sum(ws * (1-zobs)),
+           sum(ws * yobs * zobs) / sum(ws * zobs))
+  
+  gammas <- (nbk[bid,1]*(1-zobs) + nbk[bid,2]*zobs) * ws # pseudo outcomes
+  gamsbk <- list()  # s^2_b,j, sample variance
+  for (k in 1:2){
+    indk <- zobs == (k-1)
+    gammas[indk] <- gammas[indk] * (yobs[indk] - rho[k])
+    gamsbk[[k]] <- aggregate(gammas[indk], by = list(data[indk,block]), FUN = var)
+  }
+  gamsbk <- merge(gamsbk[[1]], gamsbk[[2]], by = "Group.1")[,2:3]
+  gamsbk[is.na(gamsbk)] <- 0
+  gamsb <- aggregate(gammas, by = list(bid), FUN = var)[,2] # B vector
+  
+  nu1 <- rowSums(gamsbk / nbk)
+  nu2 <- 2 / nbk[,1] / nbk[,2] * choose(nb,2) * gamsb - 
+    (1/nbk[,1] + 1/nbk[,2]) * ((nbk[,1]-1) *gamsbk[,1] + (nbk[,2]-1) *gamsbk[,2])
+  varest <- (sum(nu1[!small_blocks]) + sum(nu2[small_blocks])) / sum(data$.w0)^2
+  return(as.matrix(varest))
+}
+
+#' @title (Internal) Aggregate weights and outcomes to cluster level
+#' @param x a fitted \code{teeMod} model
+#' @details aggregate individual weights and outcomes to cluster weighted sums
+#' @return a list of 
+#' - a data frame of cluster weights, outcomes, treatments, and block ids;
+#' - treatment id column name; 
+#' - block id column name
+#' @keywords internal
+.aggregate_to_cluster <- function(x, ...){
+  ws <- if (is.null(x$weights)) 1 else x$weights
+  data_temp <- x$call$data
+  name_y <- as.character(x$terms[[2]]) # the column of y
+  data_temp <- cbind(data_temp, .w = ws, .w0 = ws / ate(x@Design, data=data_temp),
+                     .wy = ws * data_temp[, name_y])
+  
+  design_obj <- x@Design
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  
+  data_temp <- .merge_block_id_cols(data_temp, name_blk)
+  name_blk <- name_blk[1]
+  
+  eq_clu <- paste(name_clu, collapse = " + ")
+  form <- paste("cbind(", name_trt, ", ", name_blk, ") ~ ", eq_clu, sep = "")
+  form2 <- paste("cbind(.wy, .w, .w0) ~ ", eq_clu, sep = "")
+  
+  # aggregate z and bid by mean, aggregate w, w0, and w*y by sum
+  data_aggr <- cbind(
+    do.call("aggregate", list(as.formula(form), FUN = mean, data = data_temp)),
+    do.call("aggregate", list(as.formula(form2), FUN = sum, data = data_temp))
+  )
+  data_aggr$.wy <- data_aggr$.wy / data_aggr$.w
+  data_aggr <- data_aggr[order(data_aggr[, name_blk]), ]
+  return(list(data = data_aggr, block = name_blk, z = name_trt))
+}
+
+#' @title (Internal) Merge multiple block IDs
+#' @param df a data frame 
+#' @param ids a vector of block IDs, column names of df
+#' @details merge multiple block ID columns by the value combinations
+#' and store the new block ID in the column \code{ids[1]}
+#' @return a data frame with a column that contains unique block number IDs
+#' @keywords internal
+.merge_block_id_cols <- function(df, ids){
```

Comment:
I'd make sure to test the hell out of this function if you haven't already. `merge()` is maybe the least comprehensible function I've ever dealt with

Reply:
That makes sense. I've added a few more test cases (commit https://github.com/benbhansen-stats/propertee/pull/185/commits/7fbf8e2e103f0d7191d96ceb45987d08318f1aed)